### ISSUE TITLE
Endpoint path ambiguity

### ISSUE THREAD

**Issue by czack425** [oasis-tcs/openc2-impl-https#108]
There is not a definitive endpoint path within the specification other than the [examples annex](https://github.com/oasis-tcs/openc2-impl-https/blob/master/open-impl-https.md#annex-b-examples) of `/openc2` that should be receiving messages. This is likely to result in compatibility issues where varying implementers use differing paths for the same function.
The endpoint path should be defined for compatibility among implementations.

Examples:
- Implementation A, no path specified

```
POST / HTTP/1.1
Host: device_a.company.net
Content-type: application/openc2-cmd+json;version=1.0
Date: Wed, 19 Dec 2018 22:15:00 GMT
X-Request-ID: d1ac0489-ed51-4345-9175-f3078f30afe5
{	
	"action": ...
	"target": …
	"args": ...
}
```

- Implementation B, path specified

```
POST /openc2/cmd HTTP/1.1
Host: device_b.company.net
Content-type: application/openc2-cmd+json;version=1.0
Date: Wed, 19 Dec 2018 22:15:00 GMT
X-Request-ID: d1ac0489-ed51-4345-9175-f3078f30afe5
{	
	"action": ...
	"target": …
	"args": ...
}
```

**Comment by Vasileios-Mavroeidis** [comment]
I think the level of precision you require is achieved by populating the actuator specifiers in the body of the request (part of the openc2 command). This key value is not included in the example above. The consumer then has the responsibility to forward/execute the command to the relevant actuator. If you believe that we are using the headers wrongly, we missed something, or not understand your comment, please elaborate.

**Comment by czack425** [comment]
@Vasileios-Mavroeidis you misunderstood the issue.
The path for a URL is what follows the domain, it further specifies the endpoint being requested. The first example sends an HTTP POST to `http://device_a.company.net/` where as the second does the same to `http://device_b.company.net/openc2/cmd`.

Ex) The URL for this specific issue is `https://github.com/oasis-tcs/openc2-impl-https/issues/108`
- Protocol: `https://`
- Domain: `github.com`
- Path: `/oasis-tcs/openc2-impl-https/issues/108`

**Comment by dlemire60** [comment]
After giving this some thought,  I see the following approaches:

1) use the bare root with no path:  https://domain.com/
2) use an OpenC2 subdomain with no path:  https://openc2.domain.com/
2a) use an OpenC2 subdomain with an AP path:  https://openc2.domain.com/edr -- when addressing a specific AP
3) use an OpenC2 path:  https://domain.com/openc2
3a) use an OpenC2 path, with an AP extension:  https://domain.com/openc2/edr -- when addressing a specific AP

I don't know anything about how this sort of addressing might be implemented inside a private network, but that seems worth considering.  For example, rather than "domain.com" might you use a system's name as recorded in MS AD?  e.g., my machine's name is showing in Settings / System / About as "ABC123LEMIRE".  If it was running an OpenC2 consumer service, might a producer address it as https://abc123lemire/?  If so, and assuming there might be other services on the system also addressed via HTTPS, then option 3 (or 3a) seems like a more general solution.  The same would seem to apply if the system was addressed by IP address.

Assuming 3/3a as the solution, it gives us two options for directing to an AP:
   1) extend the path (option 3a)
   2) use OpenC2 message features to specify the desired AP (and/or let the receiving Consumer just "figure it out") (option 3)

Allowing both is also an option.  Option 3a is similar to what we're doing w/MQTT but I don't consider that a compelling reason to adopt it for HTTP.

**Comment by alevere** [comment]
I agree with the issue presented, there is ambiguity in the URI and it will lead to incompatible implementations. The examples show "/openc2" as path in our docs. I also recommend the #3 above such that implementors will know how to implement this and that #3 is consistent with the examples in present documentation. I think #1 and #2 would present problems for implementors who may already have a listening system with a management or api interface. I dont see creating new dns entries prefixed with openc2 as very desired or scalable. I do suggest we create a PR for this.

**Comment by dlemire60** [comment]
I was recently reminded that the IETF has an RFC that relates to this, defining `/.well-known/` as a standard prefix for predefined locations for particular services on a HTTP host.  The relevant document is [RFC 8615](https://www.rfc-editor.org/rfc/rfc8615.html). The author of an ID discussing retrieval of SBOM and vulnerability information suggested the use of the  `/.well-known/` standard for retrieval of SBOM's using OpenC2.

Perhaps OpenC2 would be better served by following the IETF's lead and using `/.well-known/openc2` as our standard path.

**Comment by alevere** [comment]
Interesting, I certainly wasnt aware of .well-known as a path. In my opinion, that isnt really any different than /openc2. I can certainly live with either, but that proposal wouldnt be backwards compatible and we would need to go through the process of registering that with IETF, and it doesnt look to me like that many orgs have chosen to go that path given the number of registrations.

**Comment by dlemire60** [comment]
* I have to agree, only 46 registered .well-known IDs over a decade (the original RFC defining the feature was in 2010) isn't very impressive. OTOH another standards org has a mechanism that fits our need so maybe it makes sense to use it.
* Backward compatible is an issue but in a documentation sense it's kind of ambiguous:  we used the path `/openc2/` in our one example and I guess people coded to that, but it was never _specified_ in document and isn't part of the conformance requirements.  The current spec only requires [listening on port 443](https://docs.oasis-open.org/openc2/open-impl-https/v1.0/cs01/open-impl-https-v1.0-cs01.html#322-http-usage).

**Comment by Vasileios-Mavroeidis** [comment]
Right, I can see a use case now from Alex's comment/perspective. 
We just need to let the consumer/listening service know that this is an openc2 command, then the next step would be to inspect the actuator specifiers. I think that Dave's suggestion number 3 is adequate. We never had this problem in our POC because we inspect the message type: ...openc2... and we forward/process it accordingly.

**Comment by dlemire60** [comment]
Resolved by PR #114. Closing issue.
